tests/test_branch.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+from contextlib import contextmanager
+from datetime import datetime, timedelta
+import os
+from pathlib import Path
+from tempfile import TemporaryDirectory
+import subprocess
+import unittest.mock
+from typing import Any, BinaryIO, Iterator
+from nix_update.version import fetch_latest_version
+from nix_update.version.version import VersionPreference
+import conftest
+
+TEST_ROOT = Path(__file__).parent.resolve()
+
+
+def create_dummy_commit(tmp: str, date: datetime) -> None:
+    datestr = date.strftime("%FT%TZ")
+    with open(Path(tmp) / "date.txt", "w", encoding="utf-8") as fp:
+        fp.write(datestr)
+        fp.write("\n")
+    subprocess.run(["git", "-C", tmp, "add", "date.txt"], check=True)
+    env = os.environ.copy()
+    env["GIT_AUTHOR_DATE"] = datestr
+    env["GIT_COMMITTER_DATE"] = datestr
+    subprocess.run(
+        [
+            "git",
+            "-C",
+            tmp,
+            "-c",
+            "user.name=dummy",
+            "-c",
+            "user.email=dummy@localhost",
+            "commit",
+            "--quiet",
+            "--message=some message",
+        ],
+        env=env,
+        check=True,
+    )
+
+
+atom_feed_log_format = """\
+  <entry>
+    <link href="/%H"/>
+    <updated>%cd</updated>
+  </entry>"""
+
+
+def create_dummy_commit_feed(tmp: str, branch: str) -> None:
+    with open(Path(tmp) / f"{branch}.atom", "wb") as feed:
+        feed.write(
+            b"""\
+<?xml version="1.0" encoding="UTF-8"?>
+<feed xmlns="http://www.w3.org/2005/Atom">
+"""
+        )
+        feed.flush()
+        env = os.environ.copy()
+        env["TZ"] = "UTC"
+        subprocess.run(
+            [
+                "git",
+                "-C",
+                tmp,
+                "log",
+                "--date=format-local:%FT%TZ",
+                f"--format=format:{atom_feed_log_format}",
+                "--max-count=20",
+                branch,
+            ],
+            env=env,
+            stdout=feed,
+            check=True,
+        )
+        feed.write(
+            b"""
+</feed>
+"""
+        )
+
+
+def fake_urlopen(url: str) -> BinaryIO:
+    feed = url.rsplit("/", 1)[-1]
+    return open(Path(os.environ["GIT_FAKE_REPO"]) / feed, "rb")
+
+
+@contextmanager
+def fake_git_repo() -> Iterator[None]:
+    """Create a temporary git repository with a couple of commits and their
+    atom feeds.
+
+    The branch master requires fetching the git repository because all the
+    commits in the feed are in the future.
+
+    The branch short's atom feed contains past commits that fetching the git
+    is not needed.
+
+    The environment variable GIT_FAKE_REPO is set to the path of this git repo.
+
+    urllib.request.urlopen is overridden to fetch our generated atom feeds.
+    """
+
+    with TemporaryDirectory(suffix=".git") as tmp:
```

Comment:
```suggestion
    with TemporaryDirectory(suffix="git") as tmp:
```
I don't like that tests could leave hidden directories behind.

Reply:
It's only the suffix so it won't be a hidden folder and unless the script is interrupted during clean-up it will be removed.

Reply:
oh, yeah true. Mixed that up.

Reply:
It's quite likely that in a test interrupt can happen i.e. when someone presses Ctrl-C. And `.git` has a special meaning, which would make things even more confusing.

Reply:
It should have some random character string before it so that shouldn't be a big problem.